### PR TITLE
SIMD.js support in Asm.js

### PR DESCRIPTION
**Pull request by nmostafa**
This change implements all SIMD.js operations in ASM.js validation, interpreter and backend.

[Review on Reviewable](https://reviewable.io/reviews/microsoft/chakracore/386)

### CODE REVIEW

**Comment by nmostafa** [comment]
This changes enables the following:
1. Support for all SIMD.js operations in ASM.js code. Validation rules are roughly described [https://github.com/tc39/ecmascript_simd/issues/307](here)
2. Enabled interpreter and lowerer/encoder to handle SIMD operations.
3. Added runtime linking checks upon module load. Change SIMD object and it is fields to configurable and writeable in Javascript Lib.
4. Cleanup: removed Float64x2, .signMask, .with{Flag}*

Missing:
1. Strict arguments for SIMD Int types to Signed variants only.

**Comment by nmostafa** [comment]
@MikeHolman, @abchatra, @LouisLaf 
Can you please review this ?

File: jenkins.check_copyright.sh [hunk 1]
```diff
@@ -32,7 +32,106 @@ git diff --name-only `git merge-base origin/master HEAD` HEAD |
...
+    grep -v -E 'test/SIMD.uint8x16.asmjs/testConversion-2.js$' |
+    grep -v -E 'test/SIMD.uint8x16.asmjs/testConversion.js$' |
+    grep -v -E 'test/SIMD.uint8x16.asmjs/testLoadStore.js$' |
+    grep -v -E 'test/SIMD.uint8x16.asmjs/testMinMax.js$' |
+    grep -v -E 'test/SIMD.uint8x16.asmjs/testMisc.js$' |
+    grep -v -E 'test/SIMD.uint8x16.asmjs/testMul.js$' |
+    grep -v -E 'test/SIMD.uint8x16.asmjs/testShift.js$' |
+    grep -v -E 'test/SIMD.uint8x16.asmjs/testShuffle.js$' |
+    grep -v -E 'test/SIMD.uint8x16.asmjs/testSplat.js$' |
+    grep -v -E 'test/SIMD.uint8x16.asmjs/testSwizzle.js$' |
```
Comment (dilijev) [inline jenkins.check_copyright.sh]: Is it possible to put all affected files in a single folder and add just that folder as an exception?

Comment (nmostafa) [inline jenkins.check_copyright.sh]: We have the tests organized by types, easier to track. I agree this is not pretty. One alternative solution is to exclude all tests (but that will exclude existing tests as well). Another is to have copyright_check be OK with Intel copyright notice only.

Comment (dilijev) [inline jenkins.check_copyright.sh]: Perhaps a subdirectory? I agree there's not really a pretty solution besides altering the check. In any case, we're still waiting to hear what Microsoft legal recommends for these cases.

**Comment by dilijev** [comment]
@dotnet-bot test this please

**Comment by MikeHolman** [comment]
Review status: 0 of 370 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[lib/Backend/IRBuilderAsmJs.cpp, line 886 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjBAWJ802goADGDB0F:-KBjBAWKSucvKCbzILAg:1092666274) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/IRBuilderAsmJs.cpp#L886)):*
personal preference would be to have case statements along lines of "argType = TySimdXX; goto simdCommon;", or maybe better yet to call a helper like GetIRTypeFromAsmVarType()

---

*[lib/Backend/IRBuilderAsmJs.cpp, line 1081 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjBnjgnkA7X-hVyVh7:-KBjBnjgnkA7X-hVyVh8:227630732) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/IRBuilderAsmJs.cpp#L1081)):*
there is currently a bunch of copy/paste code here. It feels like this will be hard to maintain. You guys might have better context, but I'd think some macro lists like `ASM_TO_IR(Bool32x4, TySimdB4, Simd128Bool32x4)`, etc. with whatever we need in a file somewhere could reduce this. Do you think such a solution could help?

---

*[lib/Backend/IRBuilderAsmJs.cpp, line 3520 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjTDvmkoPygEcOgAkn:-KBjTDvne3JhcF7jka6A:1804061178) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/IRBuilderAsmJs.cpp#L3520)):*
with so many layouts (and with all of them templates), i'm thinking this might give a non-negligible increase in binary size. can you check binary size before/after your change?

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 2463 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBk2PjZuYqxLBf_CkBA:-KBk2PjZuYqxLBf_CkBB:-447498323) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/LowerMDSharedSimd128.cpp#L2463)):*
any reason to include op encoding here? we don't really do this normally

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 2488 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBk0g5bdGw5nhYa_lWC:-KBk0g5clTEKpjOvwFg7:310918850) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/LowerMDSharedSimd128.cpp#L2488)):*
do you really need to do this? does regalloc not know to give a byteable register?

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 2608 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjl_EnqefivwbG4D8b:-KBjl_EnqefivwbG4D8c:-1928244338) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/LowerMDSharedSimd128.cpp#L2608)):*
why not do all checks up front (src < 0, src == NaN, src >= 2^32), ORPS them together and throw if nonzero? then i think you can do the `sub src, 2^31; cvt dst, src; add dst, 2^31` without needing all this masking. or is your masking method faster?

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 2655 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjpHhpXY8fSCskDKul:-KBjpHhpXY8fSCskDKum:1882306455) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/LowerMDSharedSimd128.cpp#L2655)):*
nit: comment should be `PCMPEQD`, not `CMPEQPS`

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 3481 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjenj1Ze7VCjrLmK9J:-KBjenj1Ze7VCjrLmK9K:-2145607054) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/LowerMDSharedSimd128.cpp#L3481)):*
we could put the shufMask in your tmp->GetStackSym()->scratch, and then in EncoderMD under `if (opdope & DSSE)`, read the scratch for SHUFPS, etc. That would allow for `Instr::New(Js::OpCode::SHUFPS, dst, dst, tmp, func)`, which regalloc will understand.   obviously would need to ensure all SHUFPS then use that schema. @LouisLaf, what do you think?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/386)*

**Comment by nmostafa** [comment]
Review status: 0 of 370 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[lib/Backend/IRBuilderAsmJs.cpp, line 3520 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjTDvmkoPygEcOgAkn:-KBnpIu89R_Ww6L3YYNU:-2006862515) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/IRBuilderAsmJs.cpp#L3520)):*
The increase is ~250K for x64 release. From 5.92MB to 6.17MB

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 2488 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBk0g5bdGw5nhYa_lWC:-KBnaOq9tmbpLWgyafX8:803315151) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/LowerMDSharedSimd128.cpp#L2488)):*
We ran into issues if the value goes into edi. @LouisLaf , can you advise on this ?

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 2608 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjl_EnqefivwbG4D8b:-KBnn1pNt2J-jN2HnXba:-809155207) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/LowerMDSharedSimd128.cpp#L2608)):*
I suppose it should work. However, the spec will change here to have truncation before range check. I.e. -0.5 will be in range. We can fix your approach to do src < -1 instead, but then for the CVTT we need to conditionally sub/add 2^31. Otherwise, -0.5 will become out-of-range for the CVTT. So we will have 4 SSE CMPs vs 3 CMPs we have now. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/386)*

**Comment by MikeHolman** [comment]
Reviewed 17 of 299 files at r3, 5 of 26 files at r4, 1 of 3 files at r6, 2 of 16 files at r7.
Review status: 25 of 373 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[lib/Backend/BailOut.cpp, line 727 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCIBZLT1zw3WS0DOb6Q:-KCIBZLT1zw3WS0DOb6R:1182422035) ([raw file](https://github.com/microsoft/chakracore/blob/2b1ccd069f7acfbda341ef2e31dd3bb84921ed6b/lib/Backend/BailOut.cpp#L727)):*
with all these simd checks, can was have them behind some hasSimd flag? These are cheap but we're now doing 6x more of them. And can you also measure JIT throughput?

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 2608 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjl_EnqefivwbG4D8b:-KCHjuOwDsu_DjVjaBdv:1637229834) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/LowerMDSharedSimd128.cpp#L2608)):*
Ok, that's fine with me.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 406 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCHlAd5Y577EhyanqGa:-KCHlAd6pHE3lirOgiOz:-441973649) ([raw file](https://github.com/microsoft/chakracore/blob/2b1ccd069f7acfbda341ef2e31dd3bb84921ed6b/lib/Backend/LowerMDSharedSimd128.cpp#L406)):*
nit: CMPopCode I'm reading as CM Pop Code. Could you rename to cmpOpCode?

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 745 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCHqH7MN4Mlvk1VYryu:-KCHqH7MN4Mlvk1VYryv:811733288) ([raw file](https://github.com/microsoft/chakracore/blob/2b1ccd069f7acfbda341ef2e31dd3bb84921ed6b/lib/Backend/LowerMDSharedSimd128.cpp#L745)):*
same as other comment. register allocator/legalization should be able to handle this. I believe you should only need `tmp = RegOpnd::New(TyInt8...)`, followed by `MOV tmp, dst; MOV dst, tmp`.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 773 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCHuHXN4OhN3Ztwv5m_:-KCHuHXN4OhN3Ztwv5ma:-1274151864) ([raw file](https://github.com/microsoft/chakracore/blob/2b1ccd069f7acfbda341ef2e31dd3bb84921ed6b/lib/Backend/LowerMDSharedSimd128.cpp#L773)):*
nit: you aren't pinning this to ebx, so name is confusing. Also, why not use TyInt8 to start? Generally UseWithNewType is something that should be avoided if possible.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 783 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCHv7gzGXgLxRsTJLqT:-KCHv7gzGXgLxRsTJLqU:-205883085) ([raw file](https://github.com/microsoft/chakracore/blob/2b1ccd069f7acfbda341ef2e31dd3bb84921ed6b/lib/Backend/LowerMDSharedSimd128.cpp#L783)):*
nit: it's not movd?

---

*[lib/Runtime/Language/SimdInt32x4Operation.cpp, line 7 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCILnEwDv1uRF6IPQA-:-KCILnEwDv1uRF6IPQA0:-1588984323) ([raw file](https://github.com/microsoft/chakracore/blob/1d3da4ba47a0e72b19e92a0f1a797e4d369d7d37/lib/Runtime/Language/SimdInt32x4Operation.cpp#L7)):*
for future: would be good to have this as runtime option, e.g. for non-sse2, and for testing correctness it would be useful to be able to easily compare results for scalar vs. vector implementations

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/386)*

**Comment by nmostafa** [comment]
Review status: 25 of 373 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[lib/Backend/BailOut.cpp, line 727 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCIBZLT1zw3WS0DOb6Q:-KCMdVVRhg3tdNOu-Fkr:-848463365) ([raw file](https://github.com/microsoft/chakracore/blob/2b1ccd069f7acfbda341ef2e31dd3bb84921ed6b/lib/Backend/BailOut.cpp#L727)):*
hasSimd, as if simdjs feature is enabled ? Bear in mind that this is bailout code-path, which should be rarely taken, and it's already slow. This shouldn't impact JIT throughput. 

---

*[lib/Backend/IRBuilderAsmJs.cpp, line 886 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjBAWJ802goADGDB0F:-KCMFx32-Ca5B5Zo8o7A:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/IRBuilderAsmJs.cpp#L886)):*
Done.

---

*[lib/Backend/IRBuilderAsmJs.cpp, line 1081 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjBnjgnkA7X-hVyVh7:-KCMFyz4EZuJHS4cbNBY:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/IRBuilderAsmJs.cpp#L1081)):*
Done.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 2463 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBk2PjZuYqxLBf_CkBA:-KCIe6yR49IiquYaaStI:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/LowerMDSharedSimd128.cpp#L2463)):*
Done.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 2608 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjl_EnqefivwbG4D8b:-KCMCv3cnHZclR0hBikE:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/LowerMDSharedSimd128.cpp#L2608)):*
Done.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 406 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCHlAd5Y577EhyanqGa:-KCMTlDZh2SXQRJ85s0j:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/2b1ccd069f7acfbda341ef2e31dd3bb84921ed6b/lib/Backend/LowerMDSharedSimd128.cpp#L406)):*
Done.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 745 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCHqH7MN4Mlvk1VYryu:-KCMYOe5P5OvaCTnYJ8f:499414353) ([raw file](https://github.com/microsoft/chakracore/blob/2b1ccd069f7acfbda341ef2e31dd3bb84921ed6b/lib/Backend/LowerMDSharedSimd128.cpp#L745)):*
This seems to work. I think previously, if we had a TyInt32 symbol, and we UseWithNewType as TyInt8, the regAlloc will try to cast the existing register, and if it can't (e.g. assigned to edi), we end up with 16-bit cast instead of 8-bits. 

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 773 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCHuHXN4OhN3Ztwv5m_:-KCMeE0vtfi2P-7BcqIQ:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/2b1ccd069f7acfbda341ef2e31dd3bb84921ed6b/lib/Backend/LowerMDSharedSimd128.cpp#L773)):*
Done.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 783 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCHv7gzGXgLxRsTJLqT:-KCMeDYbPBkVSwkf4Vym:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/2b1ccd069f7acfbda341ef2e31dd3bb84921ed6b/lib/Backend/LowerMDSharedSimd128.cpp#L783)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/386)*

**Comment by nmostafa** [comment]
Review status: 25 of 373 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[lib/Runtime/Language/SimdInt32x4Operation.cpp, line 7 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCILnEwDv1uRF6IPQA-:-KCMfN6gfvlUURWMGoxz:-209281265) ([raw file](https://github.com/microsoft/chakracore/blob/1d3da4ba47a0e72b19e92a0f1a797e4d369d7d37/lib/Runtime/Language/SimdInt32x4Operation.cpp#L7)):*
Agreed. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/386)*

**Comment by MikeHolman** [comment]
*[lib/Backend/BailOut.cpp, line 727 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCIBZLT1zw3WS0DOb6Q:-KCMfm5Mm24_NSXWXutD:-1454157495) ([raw file](https://github.com/microsoft/chakracore/blob/2b1ccd069f7acfbda341ef2e31dd3bb84921ed6b/lib/Backend/BailOut.cpp#L727)):*
As in, if the function has and SIMD in it. It's not just here, it looked like there were a number of places where we manipulate simd bit vectors.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/386)*

**Comment by nmostafa** [comment]
*[lib/Backend/BailOut.cpp, line 727 \[r7\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCIBZLT1zw3WS0DOb6Q:-KCMhMhendqskvpk7W4B:-271837383) ([raw file](https://github.com/microsoft/chakracore/blob/2b1ccd069f7acfbda341ef2e31dd3bb84921ed6b/lib/Backend/BailOut.cpp#L727)):*
I believe you are talking about Asmjs code then. We won't be able to tell statically otherwise. The simd bit vectors in globOpt and bailout are manipulated for type-specialization of non-Asmjs code. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/386)*

**Comment by nmostafa** [comment]
@dotnet-bot test this please

**Comment by nmostafa** [comment]
Octane perf numbers below. No perf impact.

JIT Throughput: (d:\nagy\perf\opt\x64_test\ch.exe -bgjit- -profile:backend workload.js). Lower is better.
image

Octane scores:
```
Octane            Left score       Right score      ∆ Score  ∆ Score %
----------------  ---------------  ---------------  -------  ---------
Box2d             23697.50 ±0.97%  23715.00 ±0.50%    17.50      0.07%
Code-load         14997.00 ±0.38%  14762.00 ±0.91%  -235.00     -1.57%
Crypto            18147.00 ±0.83%  18157.50 ±0.11%    10.50      0.06%
Deltablue         14382.00 ±0.84%  14503.00 ±0.48%   121.00      0.84%
Earley-boyer      24040.50 ±0.20%  24146.00 ±0.95%   105.50      0.44%
Gbemu             36171.00 ±0.05%  35802.00 ±0.49%  -369.00     -1.02%
Mandreel          18260.50 ±0.45%  18442.13 ±0.44%   181.63      0.99%
Mandreel latency  61170.50 ±0.35%  60559.75 ±1.00%  -610.75     -1.00%
Navier-stokes     27950.00 ±0.15%  27950.00 ±0.05%     0.00      0.00%
Pdfjs             15089.50 ±0.05%  15139.50 ±0.38%    50.00      0.33%
Raytrace          30711.50 ±0.99%  30774.50 ±0.89%    63.00      0.21%
Regexp             2968.00 ±0.10%   2926.00 ±0.24%   -42.00     -1.42%
Richards          15757.00 ±0.06%  15695.50 ±0.49%   -61.50     -0.39%
Splay             17574.33 ±0.43%  17700.93 ±0.39%   126.60      0.72%
Splay latency     30448.19 ±2.36%  32014.06 ±2.44%  1565.88      5.14%
Typescript        19304.00 ±0.91%  19196.00 ±0.91%  -108.00     -0.56%
Zlib              42888.50 ±0.84%  42907.00 ±0.52%    18.50      0.04%
----------------  ---------------  ---------------  -------  ---------
Total             20733.44 ±0.59%  20766.68 ±0.66%    33.24      0.16%
```

**Comment by dilijev** [comment]
@nmostafa #498 is committed to master now.

**Comment by nmostafa** [comment]
Updated PR after rebasing and changing copyright header.
@MikeHolman @dilijev please take a look.

**Comment by dilijev** [comment]
Reviewed 34 of 278 files at r9.
Review status: 34 of 372 files reviewed at latest revision, 13 unresolved discussions.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 329 \[r9\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCX40cYb1F4Tk-U_2Ru:-KCX40cYb1F4Tk-U_2Rv:-667910328) ([raw file](https://github.com/microsoft/chakracore/blob/119a007479b9cbb60ee0b523ac0be6359d474304/lib/Backend/LowerMDSharedSimd128.cpp#L329)):*
There are blank lines with trailing whitespace throughout this file.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 1421 \[r9\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCX4X_bgupxaQDt6Nd8:-KCX4X_bgupxaQDt6Nd9:1354727368) ([raw file](https://github.com/microsoft/chakracore/blob/119a007479b9cbb60ee0b523ac0be6359d474304/lib/Backend/LowerMDSharedSimd128.cpp#L1421)):*
Some instances of double blank lines throughout the file.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 2067 \[r9\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCX4RBm6oMgmltIOQSQ:-KCX4RBm6oMgmltIOQSR:1945462339) ([raw file](https://github.com/microsoft/chakracore/blob/119a007479b9cbb60ee0b523ac0be6359d474304/lib/Backend/LowerMDSharedSimd128.cpp#L2067)):*
Keep open braces on a line by themselves

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 3271 \[r9\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCX4rKELsLj_TLb_TNE:-KCX4rKELsLj_TLb_TNF:1303996657) ([raw file](https://github.com/microsoft/chakracore/blob/119a007479b9cbb60ee0b523ac0be6359d474304/lib/Backend/LowerMDSharedSimd128.cpp#L3271)):*
Some trailing spaces throughout file.

---

*[lib/Backend/Opnd.h, line 202 \[r9\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCX52P7OntChfYeH0cq:-KCX52P7OntChfYeH0cr:-381334083) ([raw file](https://github.com/microsoft/chakracore/blob/119a007479b9cbb60ee0b523ac0be6359d474304/lib/Backend/Opnd.h#L202)):*
Some inconsistency around the closing brace spacing/alignment.

---

*[lib/Backend/Sym.h, line 145 \[r9\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCX589HjqnAKq2njjcD:-KCX589HjqnAKq2njjcE:-861766075) ([raw file](https://github.com/microsoft/chakracore/blob/119a007479b9cbb60ee0b523ac0be6359d474304/lib/Backend/Sym.h#L145)):*
Inconsistent alignment here too

---

*[lib/Runtime/ByteCode/AsmJsByteCodeWriter.h, line 70 \[r9\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCX5TI2xTuryuXrDdIx:-KCX5TI3eAlvTaExElWt:1603055405) ([raw file](https://github.com/microsoft/chakracore/blob/119a007479b9cbb60ee0b523ac0be6359d474304/lib/Runtime/ByteCode/AsmJsByteCodeWriter.h#L70)):*
Trailing spaces and extra blank line

---

*[lib/Runtime/ByteCode/OpCodesSimd.h, line 313 \[r9\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCX5qbxWcXLRmm61aUx:-KCX5qbyuhIAoMVCrxc_:-1697930250) ([raw file](https://github.com/microsoft/chakracore/blob/119a007479b9cbb60ee0b523ac0be6359d474304/lib/Runtime/ByteCode/OpCodesSimd.h#L313)):*
Alignment inconsistency throughout this file

---

*[lib/Runtime/Language/AsmJsTypes.h, line 1266 \[r9\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCX6QzbGWSITdkfX5IK:-KCX6QzbGWSITdkfX5IL:2030547072) ([raw file](https://github.com/microsoft/chakracore/blob/119a007479b9cbb60ee0b523ac0be6359d474304/lib/Runtime/Language/AsmJsTypes.h#L1266)):*
trailing spaces etc

---

*[lib/Runtime/Language/AsmJsUtils.cpp, line 6 \[r9\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCX6WISiYOmgBCfDI7w:-KCX6WISiYOmgBCfDI7x:-144315873) ([raw file](https://github.com/microsoft/chakracore/blob/119a007479b9cbb60ee0b523ac0be6359d474304/lib/Runtime/Language/AsmJsUtils.cpp#L6)):*
No need for this extra line

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/386)*

**Comment by dilijev** [comment]
I can't comment on correctness of this change, but I made some comments about style.

The changes to the copyright look good. The new copyright check agrees.

**Comment by nmostafa** [comment]
Review status: 34 of 372 files reviewed at latest revision, 13 unresolved discussions.

---

*[lib/Backend/LowerMDSharedSimd128.cpp, line 3481 \[r5\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBjenj1Ze7VCjrLmK9J:-KCMk4z48ipUEg_a0wzR:-1179300885) ([raw file](https://github.com/microsoft/chakracore/blob/0e09b9f5c03a986dbc547628ed2c23e5a3b3d9f1/lib/Backend/LowerMDSharedSimd128.cpp#L3481)):*
@LouisLaf suggested we do ProcessRegUse on the  dst  if the instruction is SHUFPS (similar to if the dst is an indirOpnd the base/index are uses). 

---

*[jenkins.check_copyright.sh, line 133 \[r4\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBK4ZaY7m26_oYF6cBB-r4-133:-KCXa6Mkvh6-ggFOZKI2:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/f0fb75e63cd7bff7dff1b01279cfe3d6464fb7d6/jenkins.check_copyright.sh#L133)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/386)*

**Comment by dilijev** [comment]
Review status: 27 of 372 files reviewed at latest revision, 14 unresolved discussions.

---

*[jenkins.check_copyright.sh, line 133 \[r4\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBK4ZaY7m26_oYF6cBB-r4-133:-KCXiE4K3d4r1hEdSEkn:-691336235) ([raw file](https://github.com/microsoft/chakracore/blob/f0fb75e63cd7bff7dff1b01279cfe3d6464fb7d6/jenkins.check_copyright.sh#L133)):*
You don't need these exceptions now. Can you remove them?

---

*[jenkins.check_copyright.sh, line 134 \[r4\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCXiLgE-WMD9Rz8Chfh:-KCXiLgE-WMD9Rz8Chfi:-1453768932) ([raw file](https://github.com/microsoft/chakracore/blob/f0fb75e63cd7bff7dff1b01279cfe3d6464fb7d6/jenkins.check_copyright.sh#L134)):*
Looks like spaces got changed to a tab. This line has been updated in another PR which should be in master now, so this will conflict unless you revert.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/386)*

**Comment by nmostafa** [comment]
Review status: 27 of 372 files reviewed at latest revision, 14 unresolved discussions.

---

*[jenkins.check_copyright.sh, line 133 \[r4\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KBK4ZaY7m26_oYF6cBB-r4-133:-KCY5G631wb5rWucs9zO:1366381227) ([raw file](https://github.com/microsoft/chakracore/blob/f0fb75e63cd7bff7dff1b01279cfe3d6464fb7d6/jenkins.check_copyright.sh#L133)):*
Already removed. You were looking at an older revision.

---

*[jenkins.check_copyright.sh, line 134 \[r4\]](https://reviewable.io:443/reviews/microsoft/chakracore/386#-KCXiLgE-WMD9Rz8Chfh:-KCY58oxrFOd4Feullyd:1957721740) ([raw file](https://github.com/microsoft/chakracore/blob/f0fb75e63cd7bff7dff1b01279cfe3d6464fb7d6/jenkins.check_copyright.sh#L134)):*
I have rebased on your change and picked your version. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/386)*

**Comment by dilijev** [comment]
Reviewed 1 of 278 files at r9.
Review status: 27 of 372 files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/386)*

**Comment by MikeHolman** [comment]
LGTM

**Pull request merged**